source_integration_test.go

```diff
@@ -0,0 +1,410 @@
+}
+
+func (r Records) Less(i, j int) bool {
+	// Ensure that both records have a sort key
+	sortKeyI, okI := r[i].Key.(opencdc.StructuredData)[SortKey].(string)
+	sortKeyJ, okJ := r[j].Key.(opencdc.StructuredData)[SortKey].(string)
+
+	// If either key is not present or not a string
+	if !okI || !okJ {
+		return false
+	}
+
+	// Compare the sort key values
+	return sortKeyI < sortKeyJ
+}
+
+func (r Records) Swap(i, j int) {
+	r[i], r[j] = r[j], r[i]
+}
+
+func TestSource_SuccessfulSnapshot(t *testing.T) {
+	is := is.New(t)
+	ctx := context.Background()
+	client, cfg := prepareIntegrationTest(ctx, t)
+
+	testTable := cfg[SourceConfigTable]
+	source := &Source{}
+	err := source.Configure(context.Background(), cfg)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	// insert 5 rows
+	err = insertRecord(ctx, client, testTable, 0, 4)
+	is.NoErr(err)
+
+	err = source.Open(context.Background(), nil)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	var got Records
+	for {
+		rec, err := source.Read(ctx)
+		if errors.Is(err, sdk.ErrBackoffRetry) {
+			break
+		}
+		got = append(got, rec)
+	}
+	is.True(got != nil)
+	is.Equal(5, got.Len())
+	// sort the records then assert the values.
+	sort.Sort(got)
+	for i, rec := range got {
+		is.Equal(rec.Payload.After, opencdc.StructuredData{PartitionKey: fmt.Sprintf("pkey%d", i), SortKey: fmt.Sprintf("%d", i)})
+	}
+	_ = source.Teardown(ctx)
+}
+
+func TestSource_SnapshotRestart(t *testing.T) {
+	is := is.New(t)
+	ctx := context.Background()
+	client, cfg := prepareIntegrationTest(ctx, t)
+	testTable := cfg[SourceConfigTable]
+	source := &Source{}
+	err := source.Configure(context.Background(), cfg)
+	is.NoErr(err)
+
+	// add rows
+	err = insertRecord(ctx, client, testTable, 0, 5)
+	is.NoErr(err)
+
+	// set a non nil position
+	pos := position.Position{
+		IteratorType: position.TypeSnapshot,
+		PartitionKey: "pkey2",
+		SortKey:      "2",
+	}
+	recPos, err := pos.ToRecordPosition()
+	is.NoErr(err)
+	err = source.Open(context.Background(), recPos)
+	is.NoErr(err)
+
+	var got Records
+	for i := 0; i <= 5; i++ {
+		rec, err := source.Read(ctx)
+		is.NoErr(err)
+		got = append(got, rec)
+	}
+	// if the read records are five, then the snapshot started again successfully, from nil position
+	is.True(got != nil)
+	is.Equal(6, got.Len())
+}
+
+func TestSource_EmptyTable(t *testing.T) {
+	is := is.New(t)
+	ctx := context.Background()
+	_, cfg := prepareIntegrationTest(ctx, t)
+
+	source := &Source{}
+	err := source.Configure(context.Background(), cfg)
+	is.NoErr(err)
+	err = source.Open(context.Background(), nil)
+	is.NoErr(err)
+
+	_, err = source.Read(ctx)
+	is.True(errors.Is(err, sdk.ErrBackoffRetry))
+
+	_ = source.Teardown(ctx)
+}
+
+func TestSource_NonExistentTable(t *testing.T) {
+	is := is.New(t)
+	ctx := context.Background()
+	_, cfg := prepareIntegrationTest(ctx, t)
+
+	source := &Source{}
+
+	// set the table name to a unique uuid, so it doesn't exist.
+	cfg[SourceConfigTable] = "28e2d9e9-b1d6-4fd7-bfa1-183ceb3c9404"
```

Comment:
Any reason why we use a static string instead of doing `uuid.NewString()`?

Reply:
oh I had the table name in the error before, but after some changes it is not needed now. updated!